torch_geometric/sampler/base.py

```diff
@@ -0,0 +1,62 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, NamedTuple, Tuple, Union
+
+import torch
+
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.data.feature_store import FeatureStore
+from torch_geometric.data.graph_store import GraphStore
+from torch_geometric.typing import EdgeType, NodeType
+
+# An input to a sampler a tensor of node indices:
+SamplerInput = torch.Tensor
+
+
+# A sampler output contains the following information.
+#   * input_size: the number of input nodes to begin sampling from.
+#   * node: a tensor of output nodes resulting from sampling. In the
+#       heterogeneous case, this is a dictionary mapping node types to the
+#       associated output tensors.
+#   * row: a tensor of edge indices that correspond to the row values of the
+#       edges in the sampled subgraph. Note that these indices must be
+#       re-indexed from 0..n-1 corresponding to the nodes in the 'node' tensor.
+#       In the heterogeneous case, this is a dictionary mapping edge types to
+#       the associated output tensors.
+#   * row: a tensor of edge indices that correspond to the column values of the
+#       edges in the sampled subgraph. Note that these indices must be
+#       re-indexed from 0..n-1 corresponding to the nodes in the 'node' tensor.
+#       In the heterogeneous case, this is a dictionary mapping edge types to
+#       the associated output tensors.
+#   * edge: a tensor of the indices in the original graph; e.g. to be used to
+#       obtain edge attributes.
+class SamplerOutput(NamedTuple):
+    node: Union[torch.Tensor, Dict[NodeType, torch.Tensor]]
```

Comment:
We might want to add an optional `batch` attribute which assigns each node to an example. This will be necessary to integrate with the new `pyg-lib` implementation.

Reply:
Added a TODO.

Reply:
Done; although this may become more complicated once we try to simplify `'metadata'`. Can revert later if necessary.